xwayland/selection/selection.c

```diff
@@ -246,12 +246,65 @@ int xwm_handle_selection_event(struct wlr_xwm *xwm,
 
 void xwm_selection_init(struct wlr_xwm_selection *selection,
 		struct wlr_xwm *xwm, xcb_atom_t atom) {
-	selection->xwm = xwm;
-	selection->atom = atom;
-	selection->window = xwm->selection_window;
 	wl_list_init(&selection->incoming);
 	wl_list_init(&selection->outgoing);
 
+	selection->xwm = xwm;
+	selection->atom = atom;
+	selection->window = xcb_generate_id(xwm->xcb_conn);
+
+	if (atom == xwm->atoms[DND_SELECTION]) {
+		xcb_create_window(
+			xwm->xcb_conn,
+			XCB_COPY_FROM_PARENT,
+			selection->window,
+			xwm->screen->root,
+			0, 0,
+			8192, 8192,
+			0,
+			XCB_WINDOW_CLASS_INPUT_ONLY,
+			xwm->screen->root_visual,
+			XCB_CW_EVENT_MASK,
+			(uint32_t[]){
+				XCB_EVENT_MASK_SUBSTRUCTURE_NOTIFY | XCB_EVENT_MASK_PROPERTY_CHANGE
+			}
+		);
+
+		xcb_change_property(
+			xwm->xcb_conn,
+			XCB_PROP_MODE_REPLACE,
+			selection->window,
+			xwm->atoms[DND_AWARE],
+			XCB_ATOM_ATOM,
+			32, // format
+			1,
+			&(uint32_t){XDND_VERSION}
+		);
+	} else {
+		xcb_create_window(
+			xwm->xcb_conn,
+			XCB_COPY_FROM_PARENT,
+			selection->window,
+			xwm->screen->root,
+			0, 0,
+			10, 10,
+			0,
+			XCB_WINDOW_CLASS_INPUT_OUTPUT,
+			xwm->screen->root_visual,
+			XCB_CW_EVENT_MASK,
+			(uint32_t[]){
+				XCB_EVENT_MASK_SUBSTRUCTURE_NOTIFY | XCB_EVENT_MASK_PROPERTY_CHANGE
+			}
+		);
+
+		if (atom == xwm->atoms[CLIPBOARD]) {
+			xcb_set_selection_owner(xwm->xcb_conn, selection->window,
```

Comment:
This is copied from the pre-existing code, but as an aside: I'm not sure if this actually makes sense? 

This marks wlroots' proxy window... as a _clipboard manager for X11_?

https://www.freedesktop.org/wiki/ClipboardManager/

This seems to come with a bunch of responsibilities, like responding to `SAVE_TARGETS`  and `TARGET_SIZES` requests, none of which wlroots appears to currently do anyway.

Reply:
Yeah, I'm not sure. This is coming from Weston: https://gitlab.freedesktop.org/wayland/weston/-/commit/cba022af3d548

However the commit message doesn't explain why this has been added. Maybe it's to avoid having an X11 clipboard manager running, because it wouldn't interact properly with Wayland clients?

> This seems to come with a bunch of responsibilities, like responding to SAVE_TARGETS and TARGET_SIZES requests, none of which wlroots appears to currently do anyway.

That's no big deal, these are optional per the linked spec.

Reply:
> However the commit message doesn't explain why this has been added.

Weston also has a bit more code, which I can't find in wlroots:

https://gitlab.freedesktop.org/wayland/weston/-/commit/cba022af3d548#853f96e105430271638bdfb9a6cd981030f660a3_524_524

> Maybe it's to avoid having an X11 clipboard manager running, because it wouldn't interact properly with Wayland clients?

Hmm, maybe... I don't see why an X11 clipboard manager would break things hard enough to be explicitly disallowed (or that this even disallows one &mdash; I imagine another app can claim the `CLIPBOARD_SELECTION`?)... but I guess we can only speculate at the original decision at this point.

> That's no big deal, these are optional per the linked spec.

Right, so it's not an (obvious) problem that we grab `CLIPBOARD_MANAGER` and do nothing with it, just odd.

Reply:
FWIW, I'd be fine with dropping it and see if something breaks.